arai60_7.md

```diff
@@ -0,0 +1,196 @@
+# Step1
+- 新しく連結リストを作る or 既存の連結リストを改造して所望の連結リストを作る
+
+## 新しく連結リストを作る。
+- 連結リストをheadから順に走査して、リストに値を格納する。それを逆から読んで所望の連結リストを作ればよい。
+- 先頭のノードも作っていくので、出発点としてdummyを作っておくといい。
+```python
+class Solution:
+    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
+        visited_node_vals = []
+        current = head
+
+        while current:
+            visited_node_vals.append(current.val)
+            current = current.next
+        
+        visited_node_vals.reverse()
+        dummy = ListNode()
+        ans_node = dummy
+
+        for val in visited_node_vals:
+            ans_node.next = ListNode(val)
+            ans_node = ans_node.next 
+        
+        return dummy.next
+```
+
+- T: O(N)/ M: O(N)
+- reverse()メソッドは計算量がO(N)。
+
+## stackを使う
+- 先ほどと同じように。連結リストを走査して値をスタックに格納し取り出していけば所望の連結リストを得る。
+- ちなみに、pop()メソッドの計算量は末尾をpopするならO(1)。l[i]をpopするならO(N-i)。
+```python
+class Solution:
+    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
+        node_vals = []   #stack
+        current = head
+        
+        while current:
+            node_vals.append(current.val)
+            current = current.next
+        
+        dummy = ListNode()
+        reversed_list= dummy
+
+        while node_vals != []:
```

Comment:
`while node_vals:` と implicit false を利用したほうがシンプルになると思います。

https://google.github.io/styleguide/pyguide.html#214-truefalse-evaluations
> Use the “implicit” false if possible, e.g., if foo: rather than if foo != []:.

Reply:
Google Style Guideは一通り読むことにします。

Reply:
https://discord.com/channels/1084280443945353267/1301559648922501182/1306786530160934954
https://discord.com/channels/1084280443945353267/1301559648922501182/1306936081668767817
読むのは否定しませんがあくまでも距離感は、そういう意見もあるのか苦しゅうない、くらいでいいと思います。